### ISSUE TITLE
Impossible to explicitly call a specific method with null argument.

### ISSUE THREAD

**Issue by huxi** [spockframework/spock#589]
An issue I opened over in the Groovy JIRA was just closed as not reproducible.

[GROOVY-7845 - Impossible to explicitly call a specific method with null argument.](https://issues.apache.org/jira/browse/GROOVY-7845)

This means that the problem must be located in Spock itself.
- Clone the repository https://github.com/huxi/bugs
- cd bugs/groovy/calls-with-null
- ./gradlew

The example class has three public static 'toStringSpecial' methods. It's impossible to call toStringSpecial(Set) or toStringSpecial(String) with a null argument.

This is possible in Java by casting null to the desired type.

It's currently impossible (AFAIK) to use Spock for testing such Java methods.

Environment:

```
------------------------------------------------------------
Gradle 2.13
------------------------------------------------------------

Build time:   2016-04-25 04:10:10 UTC
Build number: none
Revision:     3b427b1481e46232107303c90be7b05079b05b1c

Groovy:       2.4.4
Ant:          Apache Ant(TM) version 1.9.6 compiled on June 29 2015
JVM:          1.8.0_92 (Oracle Corporation 25.92-b14)
OS:           Mac OS X 10.11.5 x86_64
```

**Comment by leonard84** [comment]
Yes there is a bug, but it only happens when you try to call a method with null in a `then`/`expect` block or explicit `assert`, other blocks places and places are OK. The reason is that the value recording in Spock currently casts to `Object`.

This works fine, so you can use it as a workaround for the time being:
```groovy
  def 'toStringSpecial(Set) works as expected for null.'() {
    when:
    def result = Foo.toStringSpecial((Set)null)
    
    then:
    result == 'Set: null'
  }
```

**Comment by kriegaex** [comment]
When trying to reproduce a [problem described on StackOverflow](https://stackoverflow.com/q/58279620/1082681) I was wondering why I could not reproduce it but instead the test just ran but did not react to type casts on `null` objects. Then I found this issue and realised that in Spock 1.3 (Groovy 2.5.8) it still exists. Here is how to reproduce it:

**Java application class:**

```java
package de.scrum_master.stackoverflow.q58279620;

public class ClassUnderTest {
  public static Object someMethod(String n) {
    return n == null ? "nothing" : "something";
  }

  public static Object someMethod(Integer n) {
    return n == null ? -999 : 11;
  }
}
```

**Plain Groovy class working as expected:**

This proves that it is not a Groovy but a Spock issue.

```groovy
package de.scrum_master.stackoverflow.q58279620

class Dummy {
  static void main(String[] args) {
    assert ClassUnderTest.someMethod((Integer) null) == -999
    assert ClassUnderTest.someMethod((String) null) == "nothing"
    assert ClassUnderTest.someMethod(null as Integer) == -999
    assert ClassUnderTest.someMethod(null as String) == "nothing"
  }
}
```

**Spock test showing what works and what does not:**

```groovy
package de.scrum_master.stackoverflow.q58279620

import spock.lang.Specification

class PassingNullToOverloadedMethodTest extends Specification {
  def "does not work as expected"() {
    expect:
    ClassUnderTest.someMethod((Integer) null) == -999
    ClassUnderTest.someMethod((String) null) == "nothing"
    ClassUnderTest.someMethod(null as Integer) == -999
    ClassUnderTest.someMethod(null as String) == "nothing"
  }

  def "works as expected"() {
    given:
    def result

    when:
    result = ClassUnderTest.someMethod((Integer) null)
    then:
    result == -999

    when:
    result = ClassUnderTest.someMethod((String) null)
    then:
    result == "nothing"

    when:
    result = ClassUnderTest.someMethod(null as Integer)
    then:
    result == -999

    when:
    result = ClassUnderTest.someMethod(null as String)
    then:
    result == "nothing"
  }

}
```

Please fix this ASAP. It has been open for a long time. Thank you.

**Comment by Vampire** [comment]
Fixed by #1124

**Comment by kriegaex** [comment]
I ran both Spock's `DisambiguationInConditions` smoke test and my own simple `PassingNullToOverloadedMethodTest` on the latest master and can also confirm that it works for me now. Thank you very much, @Vampire.
